tensorboard/program.py

```diff
@@ -63,125 +63,144 @@ def setup_environment():
-                   (version.VERSION, url))
-  sys.stderr.flush()
-  server.serve_forever()
-  return 0
 
-
-def launch(plugin_loaders=None, assets_zip_provider=None, **kwargs):
-  """Python API for launching TensorBoard.
-
-  Args:
-    plugin_loaders: A list of TBLoader plugin loader instances. If not
-      specified, defaults to first-party plugins.
-    assets_zip_provider: Delegates to TBContext or uses default if None.
-    kwargs: Command-line flags (e.g. logdir) as Python data structures.
-
-  Returns:
-    The URL of the TensorBoard web server, serving forever in a
-    separate thread.
-
-  Raises:
-    ValueError: If flag values are invalid.
-    socket.error: If a server could not be constructed with the host
-      and port specified. Also logs an error message.
-
-  :type plugin_loaders: list[base_plugin.TBLoader]
-  :type assets_zip_provider: () -> file
-  :rtype: str
-  """
-  # Make it easy to run TensorBoard inside other programs, e.g. Colab.
-  if plugin_loaders is None:
-    from tensorboard import default
-    plugin_loaders = default.PLUGIN_LOADERS
-  flags = _make_flags(plugin_loaders, **kwargs)
-  if assets_zip_provider is None:
-    from tensorboard import default
-    assets_zip_provider = default.get_assets_zip_provider()
-  tb_app = application.standard_tensorboard_wsgi(flags, plugin_loaders,
-                                                 assets_zip_provider)
-  server, url = make_simple_server(
-      tb_app, flags.host, flags.port, flags.path_prefix)
-  thread = threading.Thread(target=server.serve_forever, name='TensorBoard')
-  thread.daemon = True
-  thread.start()
-  return url
-
-
-def make_simple_server(tb_app, host=None, port=None, path_prefix=None):
+  def launch(self, **kwargs):
+    """Python API for launching TensorBoard.
+
+    Args:
+      kwargs: Command-line flags (e.g. logdir) as Python data structures.
+
+    Returns:
+      The URL of the TensorBoard web server, serving forever in a
+      separate thread.
+
+    Raises:
+      ValueError: If flag values are invalid.
+      socket.error: If a server could not be constructed with the host
+        and port specified. Also logs an error message.
+
+    :type plugin_loaders: list[base_plugin.TBLoader]
+    :type assets_zip_provider: () -> file
+    :rtype: str
+    """
+    # Make it easy to run TensorBoard inside other programs, e.g. Colab.
+    server, url = self._get_server(self._get_flags(**kwargs))
+    thread = threading.Thread(target=server.serve_forever, name='TensorBoard')
+    thread.daemon = True
+    thread.start()
+    return url
+
+  def _get_flags(self, **kwargs):
+    if self._flags is not None:
+      if kwargs:
+        raise ValueError('Flags already passed to constructor')
+      return self._flags
+    args = () if kwargs else sys.argv[1:]
+    parser = argparse.ArgumentParser()
+    for loader in self.plugin_loaders:
+      loader.define_flags(parser)
+    flags, unparsed = parser.parse_args(args)
```

Comment:
I think this also has to be parse_known_args(), right?  Otherwise it won't return the unparsed flags?